### PR TITLE
Add Send/Sync traits on LookupHost struct

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #22709.
cc @Veedrac

r? @bluss

I don't have added tests yet, I'll see how to do it tomorrow.

### CODE REVIEW

**Comment by bluss** [comment]
Hi! Did you find a resource that gives us confidence that this is in fact thread safe? Just wondering if you've found anything interesting.

**Comment by GuillaumeGomez** [comment]
Not yet, it was just in order to gather people. I totally missed the last discussion you had with others so this time I hope I might be included. :)

**Comment by bluss** [comment]
http://tools.ietf.org/html/rfc3493#section-6.1 says the functions are thread safe. There shouldn't be much of an issue, we just need to treat the returned pointer to linked list as an owned pointer. LookupHost is trivially Sync because we can't access anything of its internals through a &self.

**Comment by GuillaumeGomez** [comment]
Well, I guess it will be quick to check then.

**Comment by alexcrichton** [comment]
@bors: r+ 8529d75b81d20ebbf0d2d9d78a4b09bbe9d15932

Yeah reading through the docs seems to indicating nothing thread-unsafe about these functions, and regardless of the interaction between getenv, getaddrinfo, and setenv it doesn't affect the actual iterator itself.

**Comment by GuillaumeGomez** [comment]
@alexcrichton: Hold on ! Shouldn't I add tests ?

**Comment by alexcrichton** [comment]
Sure, feel free to add an amendment which adds those tests.

**Comment by nagisa** [comment]
Wow this took quite a while to happen.

**Comment by GuillaumeGomez** [comment]
@nagisa: It took me a while to find what I was looking for. ^^'

File: src/test/run-pass/sync-send-in-std.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use std::net::lookup_host;
+
+fn is_sync<T>(_: T) where T: Sync {}
+fn is_send<T>(_: T) where T: Send {}
+
+macro_rules! all_sync_send {
+    ($ctor:expr, $($iter:ident),+) => ({
+        $(
+            let mut x = $ctor;
+            is_sync(x.$iter());
```
Comment (jethrogb) [inline src/test/run-pass/sync-send-in-std.rs]: This seems wrong. You're testing whether the return value of `<LookupHost as Iterator>::next` is Send/Sync. Shouldn't you be testing `LookupHost` directly? That would be `IntoIter<SocketAddr>` now, I think.

**Pull request merged**